rest: add verbose and mempool_sequence query params for mempool/contents

andrewtoth opened the pull request:
The verbose mempool json response can get very large. This adds an option to return the non-verbose response of just the txids. It is identical to the rpc response so the diff here is minimal. This also adds the mempool_sequence parameter for rpc consistency. Verbose defaults to true to remain backwards compatible.

It uses query parameters to be compatible with the efforts in https://github.com/bitcoin/bitcoin/issues/25752.

brunoerg commented on src/rest.cpp:
```diff
@@ -608,7 +608,20 @@ static bool rest_mempool(const std::any& context, HTTPRequest* req, const std::s
+            bool verbose = raw_verbose == "true";
+            bool mempool_sequence = raw_mempool_sequence == "true";
+            if (verbose && mempool_sequence) {
+                return RESTERR(req, HTTP_BAD_REQUEST, "Verbose results cannot contain mempool sequence values. (hint: set \"verbose=false\")");
```
This error and the previous ones could be covered in the functional test

brunoerg reviewed (commented):
Concept ACK

You should update the doc (`REST-interface.md`) as well

brunoerg commented on src/rest.cpp:
```diff
@@ -608,7 +608,20 @@ static bool rest_mempool(const std::any& context, HTTPRequest* req, const std::s
         std::string str_json;
         if (param == "contents") {
-            str_json = MempoolToJSON(*mempool, true).write() + "\n";
+            std::string raw_verbose = req->GetQueryParameter("verbose").value_or("true");
```
maybe use a `const` here?
```suggestion
            const std::string raw_verbose = req->GetQueryParameter("verbose").value_or("true");
```

andrewtoth commented:
> You should update the doc (REST-interface.md) as well

@brunoerg would it make sense to update that file after there has been a release with this patch? I think that might confuse users who see those options in the doc but are unable to use them in any release.

stickies-v commented on src/rest.cpp:
```diff
@@ -608,7 +608,20 @@ static bool rest_mempool(const std::any& context, HTTPRequest* req, const std::s
         if (param == "contents") {
-            str_json = MempoolToJSON(*mempool, true).write() + "\n";
+            const std::string raw_verbose = req->GetQueryParameter("verbose").value_or("true");
+            if (raw_verbose != "true" && raw_verbose != "false") {
+                return RESTERR(req, HTTP_BAD_REQUEST, "The \"verbose\" query parameter must be either \"true\" or \"false\".");
+            }
+            const std::string raw_mempool_sequence = req->GetQueryParameter("mempool_sequence").value_or("false");
+            if (raw_mempool_sequence != "true" && raw_mempool_sequence != "false") {
+                return RESTERR(req, HTTP_BAD_REQUEST, "The \"mempool_sequence\" query parameter must be either \"true\" or \"false\".");
+            }
```
I think the most elegant solution here would be to introduce a `HTTPRequest::GetQueryParameter<bool>` specialization that converts to and returns a bool, or throws an err if the string can't be parsed. However, that would require some more fundamental changes to how we deal with errors (which unfortunately we return instead of throw) here. Flagging it here for follow-up, feel free to grab it if you're keen.

brunoerg replied:
Strong concept ACK on introducing a `HTTPRequest::GetQueryParameter<bool>`, it would be more elegant than using string ("true"/"false"), but I know it would require a lot of more work and maybe could be done in a follow-up.

andrewtoth replied:
I can definitely do that in a follow-up PR after this is merged.

stickies-v commented on src/rest.cpp:
```diff
@@ -608,7 +608,20 @@ static bool rest_mempool(const std::any& context, HTTPRequest* req, const std::s
+            if (raw_mempool_sequence != "true" && raw_mempool_sequence != "false") {
+                return RESTERR(req, HTTP_BAD_REQUEST, "The \"mempool_sequence\" query parameter must be either \"true\" or \"false\".");
+            }
+            const bool verbose = raw_verbose == "true";
```
{} list initialization is preferred (+ in other places):
```suggestion
            const bool verbose{raw_verbose == "true"};
```

stickies-v reviewed (commented):
Concept ACK - bringing it further in line with the `getrawmempool` RPC seems reasonable.

> would it make sense to update that file after there has been a release with this patch?

To minimize maintenance I'd suggest labeling this pull with the 25.0 milestone and adding a small disclaimer to the `rest-interface.md` documentation that these 2 parameters are available only from 25.0, similar to https://github.com/bitcoin/bitcoin/blob/3baa0f5a603cedd66442492346583f2b3d776db7/doc/REST-interface.md?plain=1#L56
That way when this PR doesn't make it into 25.0 it'll be easy to track and update the docs.

andrewtoth committed:
rest: add verbose and mempool_sequence query params for mempool/contents

andrewtoth committed:
tests: mempool/contents verbose and mempool_sequence query params tests

andrewtoth commented:
> I'd suggest labeling this pull with the 25.0 milestone

@stickies-v I don't think I have permissions to add labels. Otherwise I have addressed all feedback from yourself and @brunoerg .

stickies-v commented on doc/REST-interface.md:
```diff
@@ -123,11 +123,15 @@ Returns various information about the transaction mempool.
 Only supports JSON as output format.
-Refer to the `getrawmempool` RPC help for details.
+Refer to the `getrawmempool` RPC help for details. Defaults to setting 
+`verbose=true`.
```
```suggestion
`verbose=true` and `mempool_sequence=false`.
```

stickies-v commented on doc/REST-interface.md:
```diff
@@ -123,11 +123,15 @@ Returns various information about the transaction mempool.
 Refer to the `getmempoolinfo` RPC help for details.
 
-`GET /rest/mempool/contents.json`
+`GET /rest/mempool/contents.json?verbose=<true|false>&mempool_sequence=<true|false>`
```
Actually, since `false` is default, perhaps better to set that as the first option?
```suggestion
`GET /rest/mempool/contents.json?verbose=<true|false>&mempool_sequence=<false|true>`
```

stickies-v reviewed (approved):
ACK c35c7d091 (for 25.0)

> I don't think I have permissions to add labels.

The maintainers should be able to add the 25.0 label when they see this, otherwise you can always send a message on #bitcoin-core-dev IRC. It's not super urgent though, just would be a bit easier to track I think.

brunoerg commented on doc/REST-interface.md:
```diff
@@ -123,11 +123,15 @@ Returns various information about the transaction mempool.
 Only supports JSON as output format.
-Refer to the `getrawmempool` RPC help for details.
+Refer to the `getrawmempool` RPC help for details. Defaults to setting
+`verbose=true` and `mempool_sequence=false.
```
```suggestion
`verbose=true` and `mempool_sequence=false`.
```

andrewtoth replied:
Ahh sorry should have caught that.

andrewtoth committed:
doc: add mempool/contents rest verbose and mempool_sequence args

stickies-v reviewed (approved):
re-ACK [1ff5d61](https://github.com/bitcoin/bitcoin/pull/26207/commits/1ff5d61dfdaf8987e5619162662e4c760af76a43)

I verified that the only changes since c35c7d0 are suggested documentation updates to `REST-interface.md`.

pablomartin4btc reviewed (approved):
tested ACK 1ff5d61dfdaf8987e5619162662e4c760af76a43.

I've performed manual tests getting the node started with `-rest`, comparing results between the rest api call using curl with the different combinations of params and values vs. calling the `getrawtransacion` rpc command (using `bitcoin-cli`), which is essentially what the python functional tests do.

It works as expected and it's consistent with rpc.

achow101 commented:
ACK 1ff5d61dfdaf8987e5619162662e4c760af76a43

jamesmichaeltate1 commented:
(Bitcoin.org)
